File: frontend/src/types/analysis.ts
export interface BBox {
  x1: number
  y1: number
  x2: number
  y2: number
}

export interface SeatScores {
  distance: number
  occlusion: number
  angle: number
  position: number
  confidence: number
  final: number
  raw_visibility: number
}

export interface Seat {
  seat_id: string
  row_index: number
  col_index: number
  bbox: BBox
  occupied: boolean
  detection_confidence: number
  position_category: string
  manual: boolean
  scores: SeatScores
  explanation: string[]
}

export interface Instructor {
  position: [number, number]
  confidence: number
  source: 'heuristic' | 'manual'
  bbox: BBox | null
}

export interface Quality {
  resolution: string
  brightness: string
  sharpness: string
  contrast: string
  overall: number
  width: number
  height: number
  passed: boolean
  reasons: string[]
}

export interface Stats {
  seats_detected: number
  occupied: number
  empty: number
  instructor_detected: boolean
  average_visibility: number
  average_obstruction: number
}

export interface ConfidenceBreakdown {
  image_quality: number
  instructor_detection: number
  seat_detection: number
  overall: number
}

export interface AnalysisResponse {
  analysis_id: string
  quality: Quality
  instructor: Instructor | null
  instructor_needs_manual: boolean
  board_position: [number, number] | null
  seats: Seat[]
  stats: Stats
  heatmap_image: string | null
  overall_confidence: number
  confidence_breakdown: ConfidenceBreakdown
  image_width: number
  image_height: number
  detection_count: number
}

export interface BatchImageResult {
  analysis_id: string
  label: string
  analysis: AnalysisResponse
}

export interface BatchSummary {
  image_count: number
  seats_detected_min: number
  seats_detected_max: number
  most_seats_analysis_id: string | null
  note: string
}

export interface BatchResponse {
  batch_id: string
  images: BatchImageResult[]
  summary: BatchSummary
}

export type CorrectionMode = 'none' | 'move-instructor' | 'move-board' | 'add-seat'

File: backend/app/schemas/analysis.py
from __future__ import annotations

from pydantic import BaseModel


class BBoxOut(BaseModel):
    x1: float
    y1: float
    x2: float
    y2: float


class SeatScoresOut(BaseModel):
    distance: float
    occlusion: float
    angle: float
    position: float
    confidence: float
    final: float
    raw_visibility: float


class SeatOut(BaseModel):
    seat_id: str
    row_index: int
    col_index: int
    bbox: BBoxOut
    occupied: bool
    detection_confidence: float
    position_category: str
    manual: bool
    scores: SeatScoresOut
    explanation: list[str]


class InstructorOut(BaseModel):
    position: list[float]
    confidence: float
    source: str
    bbox: BBoxOut | None = None


class QualityOut(BaseModel):
    resolution: str
    brightness: str
    sharpness: str
    contrast: str
    overall: int
    width: int
    height: int
    passed: bool
    reasons: list[str]


class StatsOut(BaseModel):
    seats_detected: int
    occupied: int
    empty: int
    instructor_detected: bool
    average_visibility: float
    average_obstruction: float


class ConfidenceBreakdownOut(BaseModel):
    image_quality: float
    instructor_detection: float
    seat_detection: float
    overall: float


class AnalysisResponse(BaseModel):
    analysis_id: str
    quality: QualityOut
    instructor: InstructorOut | None
    instructor_needs_manual: bool
    board_position: list[float] | None = None
    seats: list[SeatOut]
    stats: StatsOut
    heatmap_image: str | None
    overall_confidence: float
    confidence_breakdown: ConfidenceBreakdownOut
    image_width: int
    image_height: int
    detection_count: int


class SeatCorrection(BaseModel):
    occupied: bool | None = None
    delete: bool = False


class SeatCreate(BaseModel):
    x1: float
    y1: float
    x2: float
    y2: float


class InstructorCorrection(BaseModel):
    x: float
    y: float


class BatchImageResult(BaseModel):
    analysis_id: str
    label: str
    analysis: AnalysisResponse


class BatchSummary(BaseModel):
    image_count: int
    seats_detected_min: int
    seats_detected_max: int
    most_seats_analysis_id: str | None
    note: str


class BatchResponse(BaseModel):
    batch_id: str
    images: list[BatchImageResult]
    summary: BatchSummary
